src/cosalib/meta.py

```diff
@@ -1,15 +1,86 @@
 import json
 import jsonschema
 import os.path
+import time
 
 from cosalib.builds import Builds
-from cosalib.cmdlib import write_json
+from cosalib.cmdlib import (
+    merge_dicts,
+    load_json,
+    write_json
+)
 
+COSA_DELAYED_MERGE = "coreos-assembler.delayed-meta-merge"
+COSA_VER_STAMP = "coreos-assembler.meta-stamp"
 
 SCHEMA_PATH = os.environ.get("COSA_META_SCHEMA",
                              "/usr/lib/coreos-assembler/schema/v1.json")
 
 
+class COSAMergeError(Exception):
+    """
+    Raised when unable to merge meta.json together
+    """
+    def __str__(self):
+        return f"unable to merge: {': '.join(self.args)}"
+
+
+def merge_meta(x, y):
+    """
+    Merge two GenericMeta's together
+
+    This is likely over engineered, merge_dicts is safe enough.
+    But to we want to be sure.
+
+    As a reminder: merge_dicts only merges the superset
+
+    Rules for merging:
+    - the first dict should be on-disk while second is modified one
+    - if the attribute checked exists on x, it must match on y.
+    - which ever version is newer is merged into the older
+
+    Note: this will NOT deal with replacing keys. For example,
+          re-running an AWS replication will not update the missing
+          keys.
+
+    """
+
+    # Only allow merging of artifacts from the same content
+    # i.e. you can't add RHCOS into FCOS.
+    for i in ["ostree-commit", "ostree-content-checksum",
+              "coreos-assembler.image-config-checksum"]:
+        if x.get(i, False) and (x.get(i) != y.get(i)):
+            raise COSAMergeError(f"meta.json conflict on '{i}' "
+                                 f"{x.get(i)} != {y.get(i)}")
+
+    # Do not allow merging from different COSA code.
+    cosa_git = "coreos-assembler.container-config-git"
+    x_cosa = x.get(cosa_git, {}).get("commit")
+    y_cosa = y.get(cosa_git, {}).get("commit")
+    if x_cosa and (x_cosa != y_cosa):
+        raise COSAMergeError(f"{cosa_git} version {x_cosa}"
+                             f" doest not match {y_cosa}")
+
+    ret = {}
+    x_stamp = x.get(COSA_VER_STAMP, 0)
+    y_stamp = y.get(COSA_VER_STAMP, 0)
+    if x_stamp == y_stamp:
```

Comment:
Is it actually impossible or just incredibly unlikely for two build variants to have the same stamp?

If it is indeed just incredibly unlikely then we could always just have something like so:
```python
ret = merge_dicts(*sorted([x, y], key=lambda x: x.get(COSA_VER_STAMP, 0), reverse=True))
```